### DISCUSSION TITLE
Teleop controls outside of Formant

### DISCUSSION THREAD

**Discussion by mattanimation** [polymathrobotics/caladan_examples#24] [Ideas & Requests]
There were a few times I noticed I could get stuck and realized if I hadn't had Formant up, I would have been pretty much stuck. Here are a couple of examples.

## First time
I drove into the solar field
![Screenshot from 2022-08-06 07-48-04](https://user-images.githubusercontent.com/1426997/183254464-ef914852-5148-4984-979c-498e9b792249.png)

Sent a goal to the other side of the orchard
![Screenshot from 2022-08-06 07-48-36](https://user-images.githubusercontent.com/1426997/183254465-b88e44a9-05c9-49bd-88a6-5762c98a5e44.png)

got stuck
![Screenshot from 2022-08-06 07-48-56](https://user-images.githubusercontent.com/1426997/183254466-496596e8-e5b8-47ec-ac3c-6e37ef3aa9ad.png)
![Screenshot from 2022-08-06 07-49-02](https://user-images.githubusercontent.com/1426997/183254470-360e5d32-92c0-4430-9ce0-119e4e395b7a.png)

I needed to teleop out of the local minimum.

## Second time
 So then I sent the goal again, this time when there should have been a more clear path to the goal. 

![Screenshot from 2022-08-06 07-49-58](https://user-images.githubusercontent.com/1426997/183254472-9dd7788e-a62e-4c0c-9a14-ac616be47b04.png)

indeed the path was clear, but a little too straight forward.
![Screenshot from 2022-08-06 07-51-19](https://user-images.githubusercontent.com/1426997/183254476-d4a8b9b6-4609-4c0e-846e-17ad94d8d2cd.png)

got stuck again. the lidar seemed to see something as well. :man_shrugging: :ghost: ?
![Screenshot from 2022-08-06 07-51-03](https://user-images.githubusercontent.com/1426997/183254474-832c4e19-bc3c-416c-874b-cf460691599b.png)
![Screenshot from 2022-08-06 07-51-11](https://user-images.githubusercontent.com/1426997/183254475-634ecf0d-8d50-47b4-805d-2aff9308e679.png)

I realized at this point that given the current front end in python there would have been no to to get un-stuck (from what I could tell). so I checked the API docs and noticed there were no POST commands for sending teleop... UNTIL I downloaded the `openapi.json` file and saw that there were some endpoints, but they didn't show up in the docs in Formant.

so after having fun playing with things, my suggestion would be either to add controls for manual drive to the python side using the API or at least having those be exposed in the Swagger docs.

**Comment by stefanesa** [comment]
Hey - thanks for sharing this and sorry for the delay in responding.

That's a great idea about putting those API commands into the Swagger docs.  We'll add that to our list!  Re:Python app the intention right now is to have that as simple of a demonstration as possible right now, which is why we're keeping that codebase pretty bare bones.

That being said, a great thing to build out could be a trigger in the Python app that you could click to bring up Formant's Operate view (so that you can jump into teleop when you get stuck).  That, and we'll continue to improve our planner to make it harder to get stuck!
